optuna/integration/cma.py

```diff
@@ -0,0 +1,438 @@
+            "The parameter '{}' in trial#{} is sampled independently "
+            "by using `{}` instead of `CmaEsSampler` "
+            "(optimization performance may be degraded). "
+            "You can suppress this warning by setting `warn_independent_sampling` "
+            "to `False` in the constructor of `CmaEsSampler`, "
+            "if this independent sampling is intended behavior.".format(
+                param_name, trial.number, self._independent_sampler.__class__.__name__))
+
+
+class _Optimizer(object):
+    def __init__(
+            self,
+            search_space,  # type: Dict[str, BaseDistribution]
+            x0,  # type: Dict[str, Any]
+            sigma0,  # type: float
+            cma_stds,  # type: Optional[Dict[str, float]]
+            cma_opts  # type: Dict[str, Any]
+    ):
+        # type: (...) -> None
+
+        self._search_space = search_space
+        self._param_names = list(sorted(self._search_space.keys()))
+
+        lows = []
+        highs = []
+        for param_name in self._param_names:
+            dist = self._search_space[param_name]
+            if isinstance(dist, CategoricalDistribution):
+                # Handle categorical values by ordinal representation.
+                # TODO(Yanase): Support one-hot representation.
+                lows.append(-0.5)
+                highs.append(len(dist.choices) - 0.5)
+            elif isinstance(dist, UniformDistribution) or \
+                    isinstance(dist, LogUniformDistribution):
+                lows.append(self._to_cma_params(search_space, param_name, dist.low))
+                highs.append(self._to_cma_params(search_space, param_name, dist.high))
+            elif isinstance(dist, DiscreteUniformDistribution):
+                r = dist.high - dist.low
+                lows.append(0 - 0.5 * dist.q)
+                highs.append(r + 0.5 * dist.q)
+            elif isinstance(dist, IntUniformDistribution):
+                lows.append(dist.low - 0.5)
+                highs.append(dist.high + 0.5)
+            else:
+                raise NotImplementedError('The distribution {} is not implemented.'.format(dist))
+
+        # Set initial params.
+        initial_cma_params = []
+        for param_name in self._param_names:
+            initial_cma_params.append(
+                self._to_cma_params(self._search_space, param_name, x0[param_name]))
+        cma_option = {
+            'BoundaryHandler': cma.BoundTransform,
+            'bounds': [lows, highs],
+        }
+
+        if cma_stds:
+            cma_option['CMA_stds'] = [cma_stds.get(name, 1.) for name in self._param_names]
+
+        cma_opts.update(cma_option)
+
+        self._es = cma.CMAEvolutionStrategy(initial_cma_params, sigma0, cma_opts)
+
+    def tell(self, trials, study_direction):
+        # type: (List[FrozenTrial], StudyDirection) -> int
+
+        complete_trials = []
+        for trial in trials:
+            if trial.distributions != self._search_space:
```

Comment:
This check seems too strict  (if an objective function contains any conditional parameters, the distributions of a trial never match the product search space after several trials are executed).
I think that the check can be removed because [L326](https://github.com/pfnet/optuna/pull/447/files#diff-45cb45846c0a7c8fcb996f62ffbba36cR326) verifies the compatibility of a trial.

Reply:
You're right. I fixed it with [`_collect_target_trials`](https://github.com/pfnet/optuna/pull/447/files#diff-45cb45846c0a7c8fcb996f62ffbba36cR394-R402) method.

Please note that this change also affected the following test case. Previously, the check mechanism ignored the whole trial if it contains incompatible parameters. Now, it ignores only the incompatible parameters and uses the rest of the parameters.
https://github.com/pfnet/optuna/pull/447/commits/e9f29340a41fd0c141c3f9da08ccd575bb7abab2#diff-c37400f85444aca6a57e38ab62842f67L247-R253